src/BigANNBinaryEmbeddingScenarioBase.cs

```diff
@@ -107,29 +108,35 @@ protected override ContainerProperties GetContainerSpec(string containerName)
 
         protected async Task PerformIngestion(IngestionOperationType ingestionOperationType, int? startTagId, int startVectorId, int totalVectors)
         {
-            int numIngestionBatchCount = Convert.ToInt32(this.Configurations["AppSettings:scenario:numIngestionBatchCount"]);
-            numIngestionBatchCount = (numIngestionBatchCount == 0) ? 1 : numIngestionBatchCount;
-
-            if (totalVectors % numIngestionBatchCount != 0)
-            {
-                throw new ArgumentException("Total vectors should be evenly divisible by numIngestionBatchCount");
-            }
-            int numVectorsPerRange = totalVectors / numIngestionBatchCount;
-
-            var tasks = new List<Task>(numIngestionBatchCount);
-            for (int rangeIndex = 0; rangeIndex < numIngestionBatchCount; rangeIndex++)
-            {
-                int startVectorIdForRange = startVectorId + (rangeIndex * numVectorsPerRange);
-                Console.WriteLine(
-                    $"Starting ingestion for operation {ingestionOperationType.ToString()}, range: {rangeIndex} with start vectorId: [{startVectorIdForRange}, " +
-                    $"{startVectorIdForRange + numVectorsPerRange})");
-                tasks.Add(BulkIngestDataForRange(ingestionOperationType, startTagId, startVectorIdForRange, numVectorsPerRange));
+            ConcurrentBag<int> failedIds = new ConcurrentBag<int>();
+            if (!Convert.ToBoolean(this.Configurations["AppSettings:onlyIngestFailedIds"]))
+            {
+                int numIngestionBatchCount = Convert.ToInt32(this.Configurations["AppSettings:scenario:numIngestionBatchCount"]);
+                numIngestionBatchCount = (numIngestionBatchCount == 0) ? 1 : numIngestionBatchCount;
+
+                if (totalVectors % numIngestionBatchCount != 0)
+                {
+                    throw new ArgumentException("Total vectors should be evenly divisible by numIngestionBatchCount");
+                }
+                int numVectorsPerRange = totalVectors / numIngestionBatchCount;
+
+                var tasks = new List<Task>(numIngestionBatchCount);
+
+                for (int rangeIndex = 0; rangeIndex < numIngestionBatchCount; rangeIndex++)
+                {
+                    int startVectorIdForRange = startVectorId + (rangeIndex * numVectorsPerRange);
+                    Console.WriteLine(
+                        $"Starting ingestion for operation {ingestionOperationType.ToString()}, range: {rangeIndex} with start vectorId: [{startVectorIdForRange}, " +
+                        $"{startVectorIdForRange + numVectorsPerRange})");
+                    tasks.Add(BulkIngestDataForRange(ingestionOperationType, startTagId, startVectorIdForRange, numVectorsPerRange, failedIds));
+                }
+                await Task.WhenAll(tasks);
             }
-
-            await Task.WhenAll(tasks);
+            // Retry ingestion for failed ids.
```

Comment:
This assumes retries will not fail, maybe we don't do this immeadiately after, let the operator take a look and then re-run with right flags?